source/run/src/TG4PrimaryGeneratorAction.cxx

```diff
@@ -59,128 +66,220 @@ TG4PrimaryGeneratorAction::~TG4PrimaryGeneratorAction()
-        // when event is deleted)
-        vertex = new G4PrimaryVertex(position, time);
-        event->AddPrimaryVertex(vertex);
-
-        previousVertex = vertex;
-        previousPosition = position;
-        previousTime = time;
-      }
-      else
-        vertex = previousVertex;
-
-      // Create a primary particle and add it to the vertex
-      // (primaryParticle objects are destroyed in
-      // G4EventManager::ProcessOneEvent() when event and then vertex is
-      // deleted)
-      G4ThreeVector momentum = particlesManager->GetParticleMomentum(particle);
+
+      // Current particle's momentum and energy
+      G4ThreeVector momentum = fParticlesManager->GetParticleMomentum(particle);
       G4double energy = particle->Energy() * TG4G3Units::Energy();
-      G4PrimaryParticle* primaryParticle = new G4PrimaryParticle(
-        particleDefinition, momentum.x(), momentum.y(), momentum.z(), energy);
-
-      // Set charge
-      G4double charge = particleDefinition->GetPDGCharge();
-      if (G4IonTable::IsIon(particleDefinition) &&
-          particleDefinition->GetParticleName() != "proton") {
-        // Get dynamic charge defined by user
-        TG4UserIon* userIon =
-          particlesManager->GetUserIon(particle->GetName(), false);
-        if (userIon) charge = userIon->GetQ() * eplus;
-      }
-      primaryParticle->SetCharge(charge);
 
-      // Set polarization
+      // Particle's charge,  weight and polarization
+      G4double charge = GetProperCharge(particleDefinition, particle);
+      G4double weight = particle->GetWeight();
       TVector3 polarization;
       particle->GetPolarisation(polarization);
-      primaryParticle->SetPolarization(
+      G4ThreeVector g4Polarization(
         polarization.X(), polarization.Y(), polarization.Z());
 
-      // Set weight
-      G4double weight = particle->GetWeight();
-      primaryParticle->SetWeight(weight);
+      // Create new G4PrimaryParticle and add to G4PrimaryVertex.
+      previousVertex =
+        AddParticleToVertex(event, previousVertex, particleDefinition, position,
+          time, momentum, energy, g4Polarization, charge, weight);
+    }
+  }
+}
+
+//_____________________________________________________________________________
+void TG4PrimaryGeneratorAction::TransformTracks(G4Event* event)
+{
+  /// Create a new G4PrimaryVertex objects for each TParticle
+  /// in the VMC stack.
+
+  // Cache pointers to thread-local objects
+  // The TMCManagerStack has additional info on the current track status,
+  // e.g. kinematics and geometry state.
+  // Calling TG4PrimaryGeneratorAction::TransformTracks implies that
+  // it exists, it is the same as the TVirtualMCStack above
+
+  CheckVMCStack(fMCManagerStack);
+
+  if (VerboseLevel() > 1)
+    G4cout << "TG4PrimaryGeneratorAction::TransformPrimaries: "
+           << fMCManagerStack->GetNtrack() << " particles" << G4endl;
+
+  G4PrimaryVertex* previousVertex = 0;
```

Comment:
0 -> nullptr
For consistency, it is better to use either 0 or nullptr in the same code block)